docs/spec/spec.md

```diff
@@ -0,0 +1,2314 @@
+* Connected to localhost (::1) port 9090 (#0)
+> GET /hello/openapi-doc-dygixywsw HTTP/1.1
+> Host: localhost:9090
+> User-Agent: curl/7.64.1
+> Accept: */*
+> 
+< HTTP/1.1 200 OK
+< content-type: application/json
+< content-length: 634
+< server: ballerina/2.0.0-beta.2.1
+< date: Wed, 18 Aug 2021 14:22:29 +0530
+< 
+{
+  "openapi": "3.0.1",
+  "info": {
+     "title": " hello",
+     "version": "1.0.0"
+  },
+  "servers": [
+     {
+        "url": "localhost:9090/hello"
+     }
+  ],
+  "paths": {
+     "/greeting": {
+        "get": {
+           "operationId": "operation1_get_/greeting",
+           "responses": {
+              "200": {
+                 "description": "Ok",
+                 "content": {
+                    "text/plain": {
+                       "schema": {
+                          "type": "string"
+                       }
+                    }
+                 }
+              }
+           }
+        }
+     }
+  },
+  "components": {}
+}
+```
+
+Output of OPTIONS call to service base path
+
+```ballerina
+curl -v localhost:9090/hello -X OPTIONS 
+*   Trying ::1...
+* TCP_NODELAY set
+* Connected to localhost (::1) port 9090 (#0)
+> OPTIONS /hello HTTP/1.1
+> Host: localhost:9090
+> User-Agent: curl/7.64.1
+> Accept: */*
+> 
+< HTTP/1.1 204 No Content
+< allow: GET, OPTIONS
+< link: </hello/openapi-doc-dygixywsw>;rel="service-desc"
+< server: ballerina/2.0.0-beta.2.1
+< date: Thu, 19 Aug 2021 13:47:29 +0530
+< 
+* Connection #0 to host localhost left intact
+* Closing connection 0
+```
+
+### 2.4. Client
+A client allows the program to send network messages to a remote process according to the HTTP protocol. The fixed 
+remote functions of the client object correspond to distinct network operations defined by the HTTP protocol.
+
+The client init function requires port and configuration to initialize the client. 
+```ballerina
+http:Client clientEP = check new ("http://localhost:9090", { httpVersion: "2.0" });
+```
+
+#### 2.4.1 Client types
+The client configuration can be used to enhance the client behaviour.
+
+```ballerina
+public type ClientConfiguration record {|
+    string httpVersion = HTTP_1_1;
+    ClientHttp1Settings http1Settings = {};
+    ClientHttp2Settings http2Settings = {};
+    decimal timeout = 60;
+    string forwarded = "disable";
+    FollowRedirects? followRedirects = ();
+    PoolConfiguration? poolConfig = ();
+    CacheConfig cache = {};
+    Compression compression = COMPRESSION_AUTO;
+    ClientAuthConfig? auth = ();
+    CircuitBreakerConfig? circuitBreaker = ();
+    RetryConfig? retryConfig = ();
+    CookieConfig? cookieConfig = ();
+    ResponseLimitConfigs responseLimits = {};
+    ClientSecureSocket? secureSocket = ();
+|};
+```
+
+Based on the config, the client object will be accompanied by following client behaviours. Following clients cannot be
+instantiated calling `new`, instead user have to enable the config in the `ClientConfiguration`.
+
+##### 2.4.1.1 Secure client 
```

Comment:
I wonder if the topics should be as follows,

2.4.1.1 Security
2.4.1.2 Caching
2.4.1.3 Redirect